firmware/quintuna/VC/src/app/states/app_driveState.c

```diff
@@ -1,37 +1,246 @@
+        torque_vectoring_switch_is_on = true;
+    }
+
+    if (app_canRx_CRIT_RegenSwitch_get() == SWITCH_ON)
+    {
+        app_regen_init();
+        regen_switch_is_on = true;
+    }
+}
+
+static void driveStateRunOnTick1Hz(void)
+{
+    app_allStates_runOnTick1Hz(); // NOT SURE IF QUINTUNA WILL STILL HAVE THIS 
+}
+
+static void driveStateRunOnTick100Hz(void)
+{
+    float apps_pedal_percentage  = app_canRx_FSM_PappsMappedPedalPercentage_get() * 0.01f;
+    float sapps_pedal_percentage = app_canRx_FSM_SappsMappedPedalPercentage_get() * 0.01f;
+
+    if (!driveStatePassPreCheck())
+    {
+        return;
+    }
+
+    // Disable drive buzzer after 2 seconds.
+    if (app_timer_updateAndGetState(&buzzer_timer) == TIMER_STATE_EXPIRED)
+    {
+        io_efuse_setChannel(EFUSE_CHANNEL_BUZZER, false);
+        app_canTx_VC_BuzzerOn_set(false);
+    }
+
+    // regen switched pedal percentage from [0, 100] to [0.0, 1.0] to [-0.3, 0.7] and then scaled to [-1,1]
+    if (regen_switch_is_on)
+    {
+        apps_pedal_percentage  = app_regen_pedalRemapping(apps_pedal_percentage);
+        sapps_pedal_percentage = app_regen_pedalRemapping(sapps_pedal_percentage);
+    }
+
+    app_canTx_VC_MappedPedalPercentage_set(apps_pedal_percentage);
+    runDrivingAlgorithm(apps_pedal_percentage, sapps_pedal_percentage);
 }
 
-static void driveStateRunOnTick1Hz(void) {}
+static void driveStateRunOnExit(void) {};
 
-static void driveStateRunOnTick100Hz(void) {}
+static bool driveStatePassPreCheck()
+{
+    // All states module checks for faults, and returns whether or not a fault was detected.
+    warningType warning_check = app_warningHandling_globalWarningCheck(); 
+
+    // Make sure you can only turn on VD in init and not during drive, only able to turn off
+    bool prev_regen_switch_val = regen_switch_is_on;
+    regen_switch_is_on         = app_canRx_CRIT_RegenSwitch_get() == SWITCH_ON && prev_regen_switch_val;
+
+    bool prev_torque_vectoring_switch_val = torque_vectoring_switch_is_on;
+    regen_switch_is_on = app_canRx_CRIT_TorqueVecSwitch_get() == SWITCH_ON && prev_torque_vectoring_switch_val;
+
+    /* TODO: Vehicle dyanmics people need to make sure to do a check if sensor init failed
+        or not before using closed loop features */
+    // Update Regen + TV LEDs
+    if (app_canRx_BMS_State_get() != BMS_DRIVE_STATE)
+    {
+        app_stateMachine_setNextState(&fault_state);
+    }
+
+    if (NO_WARNINGS != warning_check)
+    {
+        app_canTx_VC_INVFRTorqueSetpoint_set(INV_OFF);
+        app_canTx_VC_INVRRTorqueSetpoint_set(INV_OFF);
+        app_canTx_VC_INVFLTorqueSetpoint_set(INV_OFF);
+        app_canTx_VC_INVRLTorqueSetpoint_set(INV_OFF);
+    }
+
+    if(INVERTER_FAULT == warning_check)
+    {
 
-static void driveStateRunOnExit(void) {}
+        app_canAlerts_VC_Warning_InverterRetry_set(true);
+        app_stateMachine_setNextState(&hvInit_state);
+        // MAKE FUNCTION IN TORQUE DISTRIBUTION WHEN 4WD merged 
```

Comment:
you should think carefully about how multiple simultaneous conditions will be handled. for example, if there's an inverter fault and the start switch gets flipped in the same cycle, then `next_state` will be set to `hvInit_state` but then set again to `hv_state` 10 lines below. So inverters won't get retried in this case.

Reply:
there may be other examples of bugs like this - just one I noticed rn